server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java

```diff
@@ -379,6 +380,21 @@ public <U> SubscribableListener<U> andThen(
         return newForked(l -> addListener(l.delegateFailureAndWrap(nextStep), executor, threadContext));
     }
 
+    /**
+     * Creates and returns a new {@link SubscribableListener} {@code L} such that if this listener is completed successfully with result
+     * {@code R} then {@code fn} is invoked with argument {@code R}, and {@code L} is completed with the result of that invocation. If this
+     * listener is completed exceptionally, or {@code fn} throws an exception, then {@code L} is completed with that exception.
+     * <p>
+     * This is essentially a shorthand for a call to {@link #andThen} with a {@code nextStep} argument that is fully synchronous.
+     * <p>
+     * The threading of the {@code fn} invocation is the same as for listeners added with {@link #addListener}: if this listener is
+     * already complete then {@code fn} is invoked on the thread calling {@link #andThenCompleteWith} and in its thread context, but if this
+     * listener is incomplete then {@code fn} is invoked on the thread, and in the thread context, on which this listener is completed.
+     */
+    public <U> SubscribableListener<U> andThenCompleteWith(CheckedFunction<T, U, Exception> fn) {
```

Comment:
I've landed on preferring `andThenMap`, mostly because `map` is not available.

Reply:
I've tried that; I can live with it but I'm not totally sold on it either. WDYT about `andThenApply` here?